### ISSUE TITLE
Session doesn't have required client - Cannot redeem refresh token

### ISSUE THREAD

**Issue by dpulrichth** [keycloak/keycloak#23687]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the [latest nightly release](https://github.com/keycloak/keycloak/releases/tag/nightly)

### Area

oidc

### Describe the bug

We're running Keycloak 22 in a cluster with two nodes. We're experiencing intermittent failures when redeeming valid refresh tokens with Keycloak responding with HTTP Status Code 400:

```
{
  "error":"invalid_grant",
  "error_description":"Refresh toked issued before the client session started"
}
```

or 

```
{
  "error":"invalid_grant",
  "error_description":"Session doesn't have required client"
}
```

The latter error is also visible within the admin console which shows sessions without a client:

![image](https://github.com/keycloak/keycloak/assets/23192668/4efda817-8973-47b1-972d-18004fdb8580)

We found that we can work around this issue by disabling the "Revoke Refresh Token" setting for the affected realm. When we have this setting enabled, however, the issue appears frequently.

### Version

22.0.3

### Expected behavior

Refresh Tokens can be used.
Sessions without clients shouldn't exist (I suppose. I'm not sure what purpose a session without a client could serve)

### Actual behavior

Using refresh tokens fails intermittently.
Sessions without clients are found in Keycloak.

### How to Reproduce?

* Set up a Keycloak cluster with two nodes.
* Register a client and allow it to use refresh token
* Allow users to sign in using the client and let the client refresh tokens

Unfortunately I cannot provide a more details on how to reproduce this issue. However, it does happen quite frequently and we can see this happening by checking our logs. On bad days it happens in about one third of all attempts to redeem refresh tokens.

### Anything else?

This appears to have been around for a while. We found an issue (https://keycloak.discourse.group/t/session-lost-reference-to-the-client/15122, currently unavailable, but available in cache https://webcache.googleusercontent.com/search?q=cache:xumrWxrHm7wJ:https://keycloak.discourse.group/t/session-lost-reference-to-the-client/15122&cd=9&hl=de&ct=clnk&gl=de) which describes clients not being found in sessions as early as Keycloak 15. One user fixed this by implementing a fallback that looks up missing sessions in the database if they are not found in the cache. Perhaps the bug is related to sessions being evicted from the cache too early or caches not being in sync in a clustered setup.

There was a similar issue with offline sessions, which was also fixed by implementing a fallback to database, https://github.com/keycloak/keycloak/issues/21402

**Comment by pedroigor** [comment]
@ahus1 Can you think about any recent improvement that might be related to this issue?

**Comment by sschu** [comment]
@dpulrichth Can you confirm you don't have issues with clock synchronization? You could tweak the settings for that by allowing a certain clock skew in the realm settings.

**Comment by dpulrichth** [comment]
@sschu I couldn't find a realm setting for clock skew. The only clock skew setting I know is the one related to external identity providers which is enabled and allows up to 30 seconds of clock skew.

All session related timeouts are on default values.

I suspect this problem is somehow related to the distributed infinispan caches. I used to see this problem very frequently when running just two keycloak instances. Once I scaled up to three instances I had a lot (95%+) fewer session-related errors.

I found this in the cache documentation (https://www.keycloak.org/server/caching#_configuring_caches_for_availability) (emphasis mine)

> The default number of owners is enough to survive 1 node (owner) failure in a cluster setup with **at least three nodes**. You are free to change the number of owners accordingly to better fit into your availability requirements. To change the number of owners, open conf/cache-ispn.xml and change the value for owners=<value> for the distributed caches to your desired value.

I interpret this to mean that a Keycloak cluster MUST have at least three notes. Is this correct?

**Comment by sschu** [comment]
@dpulrichth You are right, the clock skew settings are only for external IDPs, I mixed this up.

With a distributed cache with two owners the date will be replicated to two nodes. With two nodes in the cluster, this should still survive a node failure (although after the failure the cluster is not complete until the node is back, but it should still work).

There are two things you could try to stay on two nodes a) switch to replicated caches instead of distributed caches. Their implementation is simpler so chances there is a problem with cache sync are a bit lower. b) Try to direct all traffic to only one node. This way a request should always see up-to-date data as it doesn't matter if it is synced to the other node fast enough. This doesn't fix your problem but you can at least confirm that the sync to the other node is the culprit.

**Comment by dpulrichth** [comment]
@sschu Actually I'm fine with running more than two instances so for the moment I'm sticking with three. 

If the underlying problem was the cache not being synced fast enough I would actually expect there to be more problems with three instances than with two as there's a higher chance that an instance would not be up to date. In fact, I am seeing far fewer errors - I haven't encountered any in the past seven days. Granted, this was with a low number of total sessions (~2500) but previously I encountered this with as many as one in three sessions even if there was just a handful of sessions in total.

I also didn't experience any problems with just a single instance which is as close to option b) as I can currently get. 

Option a) sounds interesting but I'd rather try to stick to Keycloak's defaults to avoid unintended side effects. 

Long story short I'll keep an eye on the cluster in the next couple of days to see if running with three nodes does in fact solve this problem. If it does I'd appreciate any pointers as to how the number of nodes may influence cache synchronization.

**Comment by ahus1** [comment]
@pedroigor - no, I am not aware of changes.

**Comment by sschu** [comment]
@dpulrichth I don't know about the internal implementation of distributed caches in Infinispan so I can't tell you what the difference is between 2 and 3 nodes. The only thing I can tell you is that it works for us with replicated caches with two nodes. And we have been running that configuration in production for years and have not seen any side effects. Quite some time ago, when we were on distributed caches, it also worked with them.
If you want to get closer to finding a cause for the distributed caches, you would have to experiment to direct traffic to single nodes only as I described.

**Comment by drohwer89** [comment]
We noticed the described issue after rolling deployments of our Keycloak instances. We were first using replicated cache for sessions and distributed cache for clientSessions. After using replicated cache for sessions and clientSessions we were not able to reproduce the issue. It seems that there is an relation between sessions and clientSessions (the following discussion suggests that: https://github.com/keycloak/keycloak/discussions/12788). 

When there is an relation between the two caches, we might should document this relation in the guides. But maybe I'm totally wrong and it's just luck that we can't reproduce the error after changing our configuration.

**Comment by sschu** [comment]
@drohwer89 There is a 1 to n relation between sessions and clientsessions. The error `Session doesn't have required client` indicates the session was found but the client session wasn't.

**Comment by mposolda** [comment]
@dpulrichth It can be also good if you add the info about your timeouts - especially SSO Session IDle timeout, SSO Session Max, Access token lifespan, client session timeouts (if you override them) as well as any timeouts overriden at client level.

**Comment by dpulrichth** [comment]
@mposolda All timeouts are Keycloak default values, I have not changed them in the realm settings or overriden for a particular clients. Current values are:

SSO Session Idle: 30 Minutes
SSO Session Max: 10 Hours
Access Token Tineout: 5Minutes
Client Session Idle / Max: Blank, should default to standard SSO Session Idle (30 Minutes)

**Comment by jnernest** [comment]
Thanks @dpulrichth disabling the "Revoke Refresh Token" option helped us. we hope this is getting fixed sooner.

**Comment by ahus1** [comment]
This issue should be resolved with the work we did for Keycloak 25 to clean up the code. Keycloak should also work better once you enable persistent sessions which are a preview feature in KC25, and which are fully supported and enabled by default in KC26.

**Comment by enastevska** [comment]
Hi @ahus1 ,

We're facing the same issue as the person who reported the bug a year ago. We're on Keycloak 25.0.5 (having the persistent user session feature enabled) and are running Keycloak on 3 pods. 

The configuration for the sessions is the following:

The realm configuration:
![Global Session Expiration](https://github.com/user-attachments/assets/4210079b-556b-421f-9d56-a25d3747a43c)

Client configuration:
![Client Session Expiration](https://github.com/user-attachments/assets/412932e5-4b91-4afb-b402-66fffc918641)
 
 We've noticed that after the session being idle for 30days the client is gone from the user's session and therefore end up getting the error. Any ideas what could be the issue?
 I tried reproducing the issue locally, just using lower values for the expiration, but I can't reproduce it.
Is there maybe some job which cleans up the clients from the sessions periodically and it's bugged? Because I've seen one scenario where the user logged in successfully once and after that they didn't take any actions for months. But the client is still somehow removed from the session, without the user triggering the refresh token call.
 
Thanks in advance.

**Comment by ahus1** [comment]
@enastevska - in general, please don't use old issue that are closed to ask questions that relate to a newer version of Keycloak. 

With the screen shots you provided, I see that the realm settings for client sessions idle time of "30 days"  is serving as a maximum when it then clears entries from the persistent sessions. In other words: You'll need to have a Client Session Max/Idle of 730 days on the realm level as well to have a client that uses that value. I would even suggest removing the client idle settings in your case altogether, as then the SSO session values are used. 

I'll open a follow-up issue that we clarify this in the docs and also add validations that align with that.

**Comment by enastevska** [comment]
@ahus1 Thanks for the fast reply, we just figured that out. Just to point out tto he code where the issue is, in the JpaUserSessionPersisterProvider#removeExpired if the "Persist user sessions" feature is enabled, then the client session idle timeout is taken from the realm configuration. Shouldn't there be another line of code to first check if there is an idle time configured on the client and that value to be considered instead?

**Comment by ahus1** [comment]
Follow-up issue created: https://github.com/keycloak/keycloak/issues/35825
